2k ram webserver

s0170071 opened the pull request:
In the current mega branch, website generation takes up to two times as much RAM as the size of the website is. If you click through the main menus, memory goes as low as 3k and that does not include the memory the core reserves then...

I completely reworked the Server.ino to directly stream out data as it is generated. There is no buffering any more. And there is no limit to the size of the website any more. 

Tested with 2.3.0, 2.4.0 and 2.4.0 git version, and LWIP 1 / 2
The amount of free memory depends on the core and LWIP version. 

2.3.0 has about 16k free
2.4.0 git and LWIP2  goes up to 18k. Lowest during TX (includes core buffer now)

Main improvement is the global StreamingBuffer class with its overloaded += operator. Usage: 

   TXBuffer.startStream();  // 200 OK Header etc.
   sendHeadandTail(F("TmplStd"),_HEAD);  // fetch the template
   TXBuffer += F("</table></form> or something else"); // content
   addFormSeparator( TXBuffer.buf); // content too.
   sendHeadandTail(F("TmplStd"),_TAIL);  // more template
  TXBuffer.endStream(); // finalize chunked stream.

For core >2.4.0 it is possible to wait for the memory to return after a sendContent command. There is almost no decrease in speed. Core 2.3.0 becomes very slow then so there is no waiting here. The memory consumption is slightly higher.

And no, I do not have chunking errors :-)
Please test, the changes were quite extensive.


![image](https://user-images.githubusercontent.com/33855380/36078903-1c284708-0f7d-11e8-9b1b-b5ccfc7dbb52.png)

s0170071 committed:
Webserver min memory

memory went as low as 3k, so wait for it to return ...

s0170071 committed:
Adjusted to min 7k free

s0170071 committed:
Timeout 100ms

s0170071 committed:
tcp_cleanup for LWIP1 & 2

s0170071 committed:
Adapted timeout

s0170071 committed:
trigger travis re-compile

trigger travis re-compile to see if the compile failure was due to a quick re-submission.

s0170071 committed:
try fix travis

s0170071 committed:
Merge remote-tracking branch 's0170071/webserverminmem' into webserverminmem

s0170071 committed:
Does Travis like extern "C" ?

s0170071 committed:
Wait for memory be like before send chunked

goes not below 9680 bytes now

s0170071 committed:
"Content-Type","text/html"

s0170071 committed:
Optimized memory usage

memory never goes below 10k now. 
Added wait for memory return on header send, 
Added wait for memory return on chunk end send,
removed duplicate of data stashing with  getWebPageTemplateVar

s0170071 committed:
Merge remote-tracking branch 's0170071/webserverminmem' into webserverminmem

s0170071 committed:
2k Webserver

s0170071 committed:
Merge remote-tracking branch 's0170071/webserverminmem' into webserverminmem

s0170071 committed:
Merge branch 'webserverminmem' into mega

TD-er commented:
Wow, and also.... woops.
Perhaps we better could have been discussing the timing of this, since there are a number of PR not yet merged which also touch the Webserver.

And merging would be a lot harder now, from v2.0 to Mega.

Let's see how this can be merged in the codebase.

TD-er commented:
Just looking at your code and you're using sendContentBlocking() and sendHeaderBlocking() within the scope of the StreamingBuffer class, which are declared outside the scope of that class.

s0170071 committed:
fix travis

s0170071 committed:
fix travis

unused variable

s0170071 committed:
Unused variable

s0170071 commented:
Arduino IDE was compiling fine all along. Travis is a bit more strict. There where two unused variables treated as errors and a missing forward declaration.

TD-er commented:
The missing forward decl. was also enough for my PlatformIO to fail building.

TD-er commented:
Just tested it on my board and it takes about 2 minutes for the ESP to become responsive (OLED framed active)
![image](https://user-images.githubusercontent.com/3751318/36086268-974216fa-0fcb-11e8-8e13-5b26c1336aeb.png)

The sendContentBlocking function does take more than 2k.

This value remains constant.

s0170071 committed:
navMenuIndex

fixed order. Needs to be set before stream starts.

s0170071 commented:
If memory goes that low, there is some website generating a giant chunk. 
This happens if too much is added to TXBuffer.buf without calling TXBuffer.checkFull.
Please check serial output (debug level) It should look like this (root page): 

```
sendcontent free: 19984 chunk size:173
sendcontent free: 19984 chunk size:584
sendcontent free: 20176 chunk size:378
sendcontent free: 20560 chunk size:397
sendcontent free: 20160 chunk size:418
sendcontent free: 20560 chunk size:394
sendcontent free: 20968 chunk size:392
sendcontent free: 21008 chunk size:183
sendcontent free: 21008 chunk size:0
```

s0170071 commented:
> Just tested it on my board and it takes about 2 minutes for the ESP to become responsive 

Please watch out for 0.0.0.0 denied ;-)

s0170071 committed:
fix download page

don't stream this one.

s0170071 committed:
Add Ram usage debug message

usually shows: 
Ram usage: Webserver only: 1048 including Core: 3120

s0170071 committed:
fix Travis

uninitialized variable.

s0170071 committed:
Travis...

TD-er commented:
> Please watch out for 0.0.0.0 denied ;-)

I disabled the IP-range check and still the unit is unpredictable unresponsive the first few minutes after boot. When the free mem gets to about 16+ kB, it gets better. But still loading the Tools page will somewhat halt the ESP.
All tested with the last commit of about an hour ago.

s0170071 commented:
Sigh. If I only could reproduce your problems I could fix them. I need more info.
What Core ? 
LWIP ? 
MHz ?
Which Devices ? tasks ? Controllers ? 
Browser ?
Can you post the serial log ?
Can you attach your config file ?

s0170071 commented:
Could this depend on the IDE ? 
If you write "somewhat halt " do you mean it is very slow ? That could be when the memory is not returning as it is supposed to and the blocking function runs into a timeout. Try to lower the time limit in 
sendContentBlocking and sendHeaderBlocking from 1000 to 50

You would probably not believe it but mine is fast as lightning, running on the latest git core and it does also run at 160MHz. But that is new- usually it crashes then. Lets see.

TD-er commented:
The almost halt part is probably due to WiFi disconnects caused by low memory.
I run the 2.3.0 Arduino core, NodeMCU with 80 MHz clock, OLED framed and a BME280 connected.

I am currently running on the 'normal' 2.0 branch again, because I was looking into the Framed plugin to get it to run with less resource requirements. (the other issue you added today ;) )
There are my settings: https://www.dropbox.com/s/ovcjco3hhkao1xb/config_ESP_EasyM_U0_Build20000_201820.dat?dl=0

s0170071 commented:
You're not running this PR on the v20 branch, are you ?

TD-er commented:
Nope, directly from your branch. Made a checkout from your branch onto my computer
With GitKraken it is very easy to switch.

s0170071 commented:
The trouble you're experiencing is connected to your config file.
This PR runs with both cores, but with your configuration it behaves like you described. I can see that now 👍 
If I disable your Domoticz MQTT controller, its running like its supposed to.
Does that ring a bell ?

s0170071 commented:
Forgot, your config runs  just fine on 2.4.0

TD-er commented:
Hmm, I was with Daniel looking into his problems with OpenHAB MQTT.
So perhaps it is bigger than only OpenHAB and maybe related to MQTT itself.

But glad you could reproduce it and that it is not me loosing my mind.

s0170071 commented:
its not running the background tasks... only after a while. 
its not getting to loop() either. Must be in setup()

s0170071 commented:
In Plugin 37 ( MQTTImport ) there is another webserver instance. Line 361 calls the connect() method. 
Is it blocking ?

TD-er commented:
This Plugin should be active to interact with MQTT controller, I hope....

TD-er commented:
But I guess we've now concluded it is very plausible it is an MQTT issue. Maybe use a central MQTT-load-thing-issue and not keep commenting on this PR?

psy0rz commented:
@s0170071 wow it sounds awesome, the memory usage was a pretty big problem! 

Another possible solution in the future would be to just serve static html/css/js files, and define a small REST api that serves meta-data and data. But thats a too big a change for now.

I'll review your changes now. (which i assume are also quite a huge change)

psy0rz commented:
look really good, we'll have to get some other PR's and development out of the way first, and some more testing, before we can merge this.

if it really seems to work well we probably will merge it to v2.0 as well, since it might solve a lot of potential ram issues. (since we're almost too short at some points)

s0170071 committed:
Cleanup and +1k Ram

- on 2.4.0 git /low mem LWIP ram goes never below 20k, during TX core dives down to 18,7k (each main menu clicked once, plain new config.)
- Fixed json output
- cleanup of send function#
- un-split css, transfer to buffer byte by byte (prevents 1.7k temporary string)
- removed commented-out function

s0170071 committed:
Merge branch 'mega' into 2kRamWebserver

s0170071 committed:
4->12 Devices per page

that was easy :-)

psy0rz commented:
ah right someone already made that configurable :)

s0170071 committed:
Fix travis

unsigned int -> int comparison

psy0rz commented:
thanks. i configured travis in such a way that all warnings are errors, to keep code clean. :)

s0170071 committed:
Update .gitignore

s0170071 committed:
Merge remote-tracking branch 'origin/mega' into 2kRamWebserver

s0170071 committed:
Merge Webserver

s0170071 committed:
Fixed String redeclaration

TD-er commented:
I have created a test build, for people to test this fix, merged with the latest Mega branch.

https://www.dropbox.com/s/sbkib3mrba5qdo2/2k_webserver_build_mega_20180220.rar?dl=0

TD-er commented:
As a test, I flashed almost all of my nodes to this build, based on this PR.
Just to see if it is stable and no other issues arise.

s0170071 commented:
@TD-er: how is your test coming ? I hear from mattlward (#908) that it seems to work on his end.

TD-er commented:
Looks like it is working as it should and also quite a bit faster.

So looks very good thus far :)
I think we should 'backport' it to v2.0

s0170071 commented:
anything I can do to support you ?

TD-er commented:
@s0170071 Write more brilliant code like this ;)

TD-er commented:
One thing you could do, is check if the recent change in Webserver.ino causes any conflicts with your branch. That would ease the merges.

s0170071 commented:
which branch do I merge where ? 
origin/mega -> s0170071/2kRamWebserver and then push to s0170071/2kRamWebserver?
or 
 s0170071/2kRamWebserver -> origin/mega and then push to s0170071/2kRamWebserver?

edit: never mind, I did it in the github web editor.

s0170071 committed:
Merge branch 'mega' into 2kRamWebserver

TD-er commented:
My idea:
create new branch: s0170071/2kRamWebserver_merge_mega
This one is then based at the head of of the current s0170071/2kRamWebserver
Then merge the current mega branch into that new branch and if it works, create a new PR for that new branch.

s0170071 commented:
replaced by #929
